### PR TITLE
Surface fusion

### PR DESCRIPTION
**Pull request by mihaibujanca**

### CODE REVIEW

**Comment by mihaibujanca** [comment]
PR quite large, perhaps should consider splitting into smaller chunks?

File: kfusion/include/kfusion/kinfu.hpp [hunk 1]
```diff
@@ -18,7 +16,10 @@ namespace kfusion
         KF_EXPORTS bool checkIfPreFermiGPU(int device);
         KF_EXPORTS void printCudaDeviceInfo(int device);
         KF_EXPORTS void printShortCudaDeviceInfo(int device);
+        class TsdfVolume;
+        class ProjectiveICP;
     }
+    class WarpField;
```
Comment (amayeh) [inline kfusion/include/kfusion/kinfu.hpp]: include header file

File: kfusion/include/kfusion/kinfu.hpp [hunk 2]
```diff
@@ -18,7 +16,10 @@ namespace kfusion
         KF_EXPORTS bool checkIfPreFermiGPU(int device);
         KF_EXPORTS void printCudaDeviceInfo(int device);
         KF_EXPORTS void printShortCudaDeviceInfo(int device);
+        class TsdfVolume;
```
Comment (amayeh) [inline kfusion/include/kfusion/kinfu.hpp]: include header file

File: kfusion/include/kfusion/kinfu.hpp [hunk 3]
```diff
@@ -18,7 +16,10 @@ namespace kfusion
...
+        class ProjectiveICP;
```
Comment (amayeh) [inline kfusion/include/kfusion/kinfu.hpp]: include header file

File: kfusion/include/kfusion/warp_field.hpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+ */
+#include <dual_quaternion.hpp>
+#include <kfusion/types.hpp>
+#include <nanoflann/nanoflann.hpp>
+
+namespace kfusion
+{
+    namespace cuda
+    {
+        class TsdfVolume;
```
Comment (amayeh) [inline kfusion/include/kfusion/warp_field.hpp]: include header file

File: kfusion/include/kfusion/warp_field.hpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    }
+    namespace utils
+    {
+        class PointCloud;
```
Comment (amayeh) [inline kfusion/include/kfusion/warp_field.hpp]: include header file

File: kfusion/include/kfusion/warp_field.hpp [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+     * \var node::index
+     * Index of the node in the canonical frame. Equivalent to dg_v
+     *
+     * \var node::transform
+     * Translation and rotation of a node, equivalent to dg_se in the paper
+     *
+     * \var node::weight
+     * Equivalent to dg_w
+     */
+    struct node
```
Comment (amayeh) [inline kfusion/include/kfusion/warp_field.hpp]: plz use better naming

File: kfusion/include/kfusion/warp_field.hpp [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            return std::abs(x) <= c ? x * std::pow((1 - (x * x) / (c * c)), 2) : 0.0;
+        }
+
+        /**
+         * Huber penalty function, implemented as described in https://en.wikipedia.org/wiki/Huber_loss
+         * \param a
+         * \param delta
+         * \return
+         */
+        inline float huberPenalty(float a, float delta) const
```
Comment (amayeh) [inline kfusion/include/kfusion/warp_field.hpp]: plz move it to cpp file

File: kfusion/include/kfusion/warp_field.hpp [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+         * \param x
+         * \param c
+         * \return
+         * 
+         * \note 
+         * The value c = 4.685 is usually used for this loss function, and
+         * it provides an asymptotic efficiency 95% that of linear
+         * regression for the normal distribution
+         */
+        inline float tukeyPenalty(float x, float c) const
```
Comment (amayeh) [inline kfusion/include/kfusion/warp_field.hpp]: plz move it to cpp

**Comment by zzhang001** [comment]
What's the run instruction per our discussion? Please also send the status shared doc link again to me. Thanks!

**Comment by zzhang001** [comment]
@mihaibujanca We have docs, surface fusion, quaternion test three branches. Can we please merge them all into master? Thanks!

**Pull request merged**